File: solutions/cpp/roman-numerals/1/roman_numerals.cpp
#include "roman_numerals.h"
#include <iostream>

using std::string;

// Function to convert an integer to its Roman numeral representation
namespace roman_numerals {
  string convert(int n) {

    // this algorithm needs the array to be from greatest to least
    string str_romans[] = {"M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"};
    int values[] = {1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1};
    int len = sizeof(values)/sizeof(values[0]); 

    // Initialize an empty string to hold the resulting Roman numeral
    string result = "";

    // Loop through the values and corresponding Roman symbols
    for (auto i = 0; i < len; ++i)
    {
        // While the current number is greater than or equal to the value
        while(n - values[i] >= 0)
        {
            // Append the Roman symbol to the result and subtract the value from the number
            result += str_romans[i];
            n -= values[i];
        }
    }

    // Return the resulting Roman numeral
    return result;
  }

}  // namespace roman_numerals

File: solutions/elixir/roman-numerals/1/lib/roman_numerals.ex
defmodule RomanNumerals do

  @roman_steps [
    {1000, "M"},
    {900, "CM"},
    {500, "D"},
    {400, "CD"},
    {100, "C"},
    {90, "XC"},
    {50, "L"},
    {40, "XL"},
    {10, "X"},
    {9, "IX"},
    {5, "V"},
    {4, "IV"},
    {1, "I"}
  ]

  @doc """
  Convert the number to a roman number.
  """
  @spec numeral(pos_integer) :: String.t()
  def numeral(0), do: ""
  def numeral(number) when number > 0 do
    {base, roman} = Enum.find(@roman_steps, fn {base, _} -> number >= base end)
    roman <> numeral(number - base)
  end
end
